cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -52,57 +54,283 @@ func ValidateMasterConfiguration(c *kubeadm.MasterConfiguration) field.ErrorList
 
 func ValidateNodeConfiguration(c *kubeadm.NodeConfiguration) field.ErrorList {
 	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, ValidateDiscovery(&c.Discovery, field.NewPath("discovery"))...)
+	allErrs = append(allErrs, ValidateDiscovery(c, field.NewPath("discovery"))...)
 	return allErrs
 }
 
-func ValidateDiscovery(c *kubeadm.Discovery, fldPath *field.Path) field.ErrorList {
+func ValidateAuthorizationMode(authzMode string, fldPath *field.Path) field.ErrorList {
+	if !authzmodes.IsValidAuthorizationMode(authzMode) {
+		return field.ErrorList{field.Invalid(fldPath, nil, "invalid authorization mode")}
+	}
+	return field.ErrorList{}
+}
+
+func ValidateDiscovery(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidateArgSelection(c, fldPath)...)
 	var count int
-	if c.Token != nil {
-		allErrs = append(allErrs, ValidateTokenDiscovery(c.Token, fldPath)...)
+	if len(c.DiscoveryToken) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryToken(c, fldPath)...)
 		count++
 	}
-	if c.File != nil {
-		allErrs = append(allErrs, ValidateFileDiscovery(c.File, fldPath)...)
+	if len(c.DiscoveryFile) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryFile(c, fldPath)...)
 		count++
 	}
-	if c.HTTPS != nil {
-		allErrs = append(allErrs, ValidateHTTPSDiscovery(c.HTTPS, fldPath)...)
+	if len(c.DiscoveryURL) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryURL(c, fldPath)...)
 		count++
 	}
-	if count != 1 {
+	allErrs = append(allErrs, ValidateTLSBootstrapToken(c, fldPath)...)
+	allErrs = append(allErrs, ValidateJoinMastersArgs(c, fldPath)...)
+	if count > 1 {
 		allErrs = append(allErrs, field.Invalid(fldPath, nil, "exactly one discovery strategy can be provided"))
 	}
 	return allErrs
 }
 
-func ValidateFileDiscovery(c *kubeadm.FileDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateArgSelection(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidateTokenArg(cfg, field.NewPath("discovery"))...)
+	allErrs = append(allErrs, ValidateDiscoveryTokenArg(cfg, field.NewPath("discovery"))...)
+	allErrs = append(allErrs, ValidateTLSBootstrapTokenArg(cfg, field.NewPath("discovery"))...)
+	allErrs = append(allErrs, ValidateDiscoveryFileArg(cfg, field.NewPath("discovery"))...)
+	allErrs = append(allErrs, ValidateDiscoveryURLArg(cfg, field.NewPath("discovery"))...)
+	if len(cfg.Token) != 0 {
```

Comment:
I'm 99.9% sure that defaulting is done before validation (if not, then it's a critical bug), so I think we can move this to the defaulting code and say:
```go
if len(cfg.Token) != 0 && len(cfg.TLSBootstrapToken) == 0 && len(cfg.DiscoveryToken) == 0 {
    cfg.TLSBootstrapToken = cfg.Token
    cfg.DiscoveryToken = cfg.Token
}
```